telegram/utils/helpers.py

```diff
@@ -113,3 +120,45 @@ def mention_markdown(user_id, name):
     """
     if isinstance(user_id, int):
         return '[{}](tg://user?id={})'.format(escape_markdown(name), user_id)
+
+
+def _extract_urls_from_text(text):
+    """
+    Returns a list of urls from a text string.
+    URLs without a leading `http://` or `www.` won't be found.
+    """
+    out = []
+    for word in text.split(' '):
+        thing = urlparse(word.strip())
+        if thing.scheme:
+            out.append(word)
+    return out
+
+
+def extract_urls(message):
+    """
+    Extracts all Hyperlinks that are contained in a message. This includes
+    message entities and the media caption. Distinct links are returned in order of appearance.
+
+    Note: Exact duplicates are removed, but there may still be URLs that link
+    to the same resource.
+
+    Args:
+        message (:obj:`telegram.Message`) The message to extract from
+
+    Returns:
+        :obj:`list`: A list of URLs contained in the message
+    """
+    from telegram import MessageEntity
+
+    types = [MessageEntity.URL, MessageEntity.TEXT_LINK]
+    results = message.parse_entities(types=types)
+    results.update(message.parse_caption_entities(types=types))
+    all_urls = [v if k.type == MessageEntity.URL else k.url for k, v in results.items()]
```

Comment:
Please use a generator expression instead of a list comprehension.
i.e.
`all_urls = (v if k.type ...)`

Reply:
Is this because it saves memory by not building up the whole list but only one element at a time?